### PR TITLE
Transformer forward

### PR DESCRIPTION
**Pull request by riveSunder**

### CODE REVIEW

File: coverage.txt [hunk 1]
```diff
@@ -6,7 +6,8 @@ sepia/graph/__init__.py         0      0   100%
 sepia/graph/data.py            13      7    46%
 sepia/graph/functional.py      26     17    35%
 sepia/seq/__init__.py           0      0   100%
-sepia/seq/data.py              36     27    25%
+sepia/seq/data.py              36     11    69%
 sepia/seq/functional.py        87      0   100%
+sepia/seq/transformer.py      138     13    91%
 -----------------------------------------------
-TOTAL                         166     51    69%
+TOTAL                         304     48    84%
```
Comment (riveSunder) [inline coverage.txt:13]: ✓

File: sepia/seq/functional.py [hunk 2]
```diff
@@ -22,18 +22,18 @@
 DecoderParams = namedtuple("DecoderParams", \
         field_names=("encoded_attention", "mlp_params"))
 MLPParams = namedtuple("MLPParams", \
-        field_names=("mlp_weights", "mlp_biases", "activation"))
+        field_names=("mlp_weights", "mlp_biases"))
```
Comment (riveSunder) [inline sepia/seq/functional.py:25]: Removed activation. the activation field was (unsurprisingly) causing problems for `jax.grad`

File: sepia/seq/transformer.py [hunk 3]
```diff
@@ -54,14 +54,16 @@ def __init__(self, **kwargs):
         self.encoder_dim = self.token_dim * 3
 
         # expose these to user
-        self.encoder_size = 1
-        self.decoder_size = 1
+        self.seq_length = 32
+        self.encoder_size = query_kwargs("encoder_size", 4, **kwargs)
```
Comment (riveSunder) [inline sepia/seq/transformer.py:58]: Every hand-coded parameter in this section is due for exposure to user choice with `**kwargs`

File: sepia/seq/transformer.py [hunk 4]
```diff
@@ -78,22 +80,20 @@ def initialize_model(self):
...
         self.encoder_stack = []
         for ii in range(self.encoder_size): 
-            weights = npr.randn(self.token_dim, self.encoder_dim)
+            weights = npr.randn(self.token_dim, self.encoder_dim)*self.init_scale
             attention_weights = SelfAttentionW(weights = weights)
 
-            mlp_weights = [npr.randn(self.token_dim, self.mlp_hidden_dim), \
-                    npr.randn(self.mlp_hidden_dim, self.mlp_hidden_dim), \
-                    npr.randn(self.mlp_hidden_dim, self.token_dim)]
+            mlp_weights = [npr.randn(self.token_dim, self.mlp_hidden_dim)*self.init_scale, \
```
Comment (riveSunder) [inline sepia/seq/transformer.py:86]: In future development we should implement stand-alone initialization functions.

File: sepia/seq/transformer.py [hunk 5]
```diff
@@ -178,20 +178,195 @@ def __call__(self, sequence: str) -> str:
...
+                self.encoder_stack, \
+                self.decoder_stack]
+
+
+        masked_tokens = input_tokens \
+                * (npr.rand(*input_tokens.shape[:2],1) > self.mask_rate)
+
+        grad_loss = grad(self.get_loss, argnums=2)
+
+        # splitting these roles (returning loss or returning grads) might speed up training
```
Comment (riveSunder) [inline sepia/seq/transformer.py:221]: This may be a good place to improve execution speed, we don't necessarily need to calculate loss every time, or calculate loss and grads separately.

File: sepia/seq/transformer.py [hunk 6]
```diff
@@ -178,20 +178,195 @@ def __call__(self, sequence: str) -> str:
...
+        loss = self.get_loss(masked_tokens, input_tokens, parameters)
+        my_grad = grad_loss(masked_tokens, input_tokens, parameters)
+
+        self.parameters = parameters
+
+        return loss, my_grad
+
+    def optimizer_step(self, parameters, gradients):
```
Comment (riveSunder) [inline sepia/seq/transformer.py:229]: Future plans include more options for different types of optimizers (_e.g._ Adam)

File: sepia/seq/transformer.py [hunk 7]
```diff
@@ -178,20 +178,195 @@ def __call__(self, sequence: str) -> str:
...
+
+        """
+        (token_parameters, encoder_stack, decoder_stack)
+        
+        token_parameters: 
+            NICEParametersW = namedtuple("NICEParametersW", field_names=("weights"))
```
Comment (riveSunder) [inline sepia/seq/transformer.py:235]: The `nametuple`s used for parameters can be simplified. For example, the named tuple `DecoderParams`  should probably just have three field names for `mlp_params`, `self_attention`, and `encoder_decoder_attention` instead of using the somewhat redundant `EncodedAttentionW`.

File: sepia/seq/transformer.py [hunk 8]
```diff
@@ -178,20 +178,195 @@ def __call__(self, sequence: str) -> str:
...
+                        field_names=("self_weights", "encoded_weights")) 
+                which includes as self_weights, encoded_weights:
+                    SelfAttentionW = namedtuple("SelfAttentionW", field_names=("weights"))
+                    SelfAttentionW = namedtuple("SelfAttentionW", field_names=("weights"))
+                MLPParams = namedtuple("MLPParams", \
+                        field_names=("mlp_weights", "mlp_biases"))
+
+        """
+        
+        # SGD
```
Comment (riveSunder) [inline sepia/seq/transformer.py:255]: Also, we should make optimizer steps more general, instead of the specific routine written for `Transformer` here.

File: sepia/seq/transformer.py [hunk 9]
```diff
@@ -178,20 +178,195 @@ def __call__(self, sequence: str) -> str:
...
+                        field_names=("mlp_weights", "mlp_biases"))
+        decoder_stackk:
+            [DecoderParams = namedtuple("DecoderParams", field_names=("encoded_attention", "mlp_params"))]
+            which includes as encoded_attention, mlp_params:
+                EncodedAttentionW = namedtuple("EncodedAttentionW", \
+                        field_names=("self_weights", "encoded_weights")) 
+                which includes as self_weights, encoded_weights:
+                    SelfAttentionW = namedtuple("SelfAttentionW", field_names=("weights"))
+                    SelfAttentionW = namedtuple("SelfAttentionW", field_names=("weights"))
+                MLPParams = namedtuple("MLPParams", \
```
Comment (riveSunder) [inline sepia/seq/transformer.py:250]: This comment was used as reference when implementing the function, but should be removed.

**Review by riveSunder — COMMENTED** [review]
Remove reference comment/dead code in `optimizer_step`
Then ok to merge.

File: sepia/seq/transformer.py (hunk 9 above)
Comment (riveSunder) [reply sepia/seq/transformer.py]: commit c8b2bb1a7abb06cb1696a44c89839e2a8a8f1398
